File: sdk/js/src/RTC.js
import { EventEmitter } from 'events';

const ices = 'stun:stun.stunprotocol.org:3478'

export default class RTC extends EventEmitter {
    constructor() {
        super();
        this._sender = {};
        this._receivers = new Map();
    }

    get sender () {
        return this._sender;
    }

    getReceivers (pubid) {
        return this._receivers.get(pubid);
    }

    async createSender(pubid) {
        let sender = {
            offerSent: false,
            pc: null,
        };
        sender.pc = new RTCPeerConnection({ iceServers: [{ urls: ices }] });
        let stream = await navigator.mediaDevices.getUserMedia({ video: true, audio: true });
        sender.pc.addStream(stream);
        this.emit('localstream', pubid, stream);
        this._sender = sender;
        return sender;
    }

    createRecver (pubid) {
        try {
            let receiver = {
                offerSent: false,
                pc: null,
                id: pubid,
                streams: []
            };

            var pc = new RTCPeerConnection({ iceServers: [{ urls: ices }] });
            pc.onicecandidate = e => {
                console.log('receiver.pc.onicecandidate => ' + e.candidate)
            }

            pc.addTransceiver('audio', { 'direction': 'recvonly' });
            pc.addTransceiver('video', { 'direction': 'recvonly' });

            pc.onaddstream = (e) => {
                var stream = e.stream;
                console.log('receiver.pc.onaddstream', stream.id);
                var receiver = this._receivers.get(pubid);
                receiver.streams.push(stream);
                this.emit('addstream', pubid, stream);
            }

            pc.onremovestream = (e) => {
                var stream = e.stream;
                console.log('receiver.pc.onremovestream', stream.id);
                this.emit('removestream', pubid, stream);
            }
            receiver.pc = pc;
            this._receivers.set(pubid,receiver);
            return receiver;
        } catch (e) {
            console.log(e);
            throw e;
        }
    }

    closeRecver(pubid) {
        var receiver = this._receivers.get(pubid)
        if(receiver) {
            receiver.streams.forEach(stream => {
                this.emit('removestream', pubid, stream)
            })
            receiver.pc.close();
            this._receivers.delete(pubid);
        }
    }
}

File: sdk/flutter/lib/src/rtc.dart
import 'package:events2/events2.dart';
import 'package:flutter_webrtc/webrtc.dart';
import 'logger.dart' show Logger;

class Sender {
  Sender(this.pc, this.stream, this.id);
  String id;
  RTCPeerConnection pc;
  bool senderOffer = false;
  MediaStream stream;
}

class Receiver {
  Receiver(this.publid);
  RTCPeerConnection pc;
  bool senderOffer = false;
  String publid;
  List<MediaStream> streams = new List();
}

final Map<String, dynamic> configuration = {
    "iceServers": [
      {"url": "stun:stun.stunprotocol.org:3478"},
    ],
    'sdpSemantics': 'plan-b',
  };

final Map<String, dynamic> mediaConstraints = {
      "audio": true,
      "video": {
        "mandatory": {
          "minWidth": '640',
          "minHeight": '480',
          "minFrameRate": '30',
        },
        "facingMode": "user",
        "optional": [],
      }
    };

class RTC extends EventEmitter {
  var logger = new Logger("Pion::RTC");
  Sender _sender;
  Map<String, Receiver> _receivers = new Map();

  RTC();

  get sender => _sender;

  get receivers => _receivers;

  Future<Sender> createSender(var pubid) async {
        final Map<String, dynamic> constraints = {
          'mandatory': {
            'OfferToReceiveAudio': false,
            'OfferToReceiveVideo': false,
          },
          'optional': [],
        };
        var pc = await createPeerConnection(configuration, constraints);
        var stream = await navigator.getUserMedia(mediaConstraints);
        pc.addStream(stream);
        this.emit('localstream', pubid, stream);
        _sender = Sender(pc, stream,pubid);
        return _sender;
    }

  Future<Receiver> createRecver(pubid) async {
      final Map<String, dynamic> _constraints = {
          'mandatory': {
            'OfferToReceiveAudio': true,
            'OfferToReceiveVideo': true,
          },
          'optional': [],
        };
        try {
            var receiver = Receiver(pubid);
            RTCPeerConnection pc = await createPeerConnection(configuration, _constraints);
            pc.onAddStream = (stream){
                logger.debug('receiver.pc.onaddstream = ' + stream.id);
                var receiver = _receivers[pubid];
                receiver.streams.add(stream);
                this.emit('addstream', pubid, stream);
            };
            pc.onRemoveStream = (stream) {
                logger.debug('receiver.pc.onremovestream = ' + stream.id);
                _receivers.remove(pubid);
                this.emit('removestream', pubid, stream);
            };
            receiver.pc = pc;
            _receivers[pubid] = receiver;
            return receiver;
        } catch (e) {
            logger.debug(e);
            throw e;
        }
    }

    closeSender() async {
      if(_sender != null){
        this.emit('removelocalstream', _sender.id, _sender.stream);
        await _sender.stream.dispose();
        await _sender.pc.close();
        _sender = null;
      }
    }

    closeReceiver(pubid) {
        var receiver = _receivers[pubid];
        if(receiver != null) {
            receiver.streams.forEach((stream) {
                this.emit('removestream', pubid, stream);
            });
            receiver.pc.close();
            _receivers.remove(pubid);
        }
    }

    closeReceivers() async {
      _receivers.forEach((pubid, receiver) async {
            receiver.streams.forEach((stream) async {
                await stream.dispose();
                this.emit('removestream', pubid, stream);
            });
            await receiver.pc.close();
      });
    }
}
